google/cloud/pubsub/internal/tracing_exactly_once_ack_handler.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsub/exactly_once_ack_handler.h"
+#include "google/cloud/pubsub/internal/span.h"
+#include "google/cloud/pubsub/internal/tracing_helpers.h"
+#include "google/cloud/pubsub/version.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/status.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <opentelemetry/context/runtime_context.h>
+#include <opentelemetry/trace/context.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+#include <opentelemetry/trace/span.h>
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <cstdint>
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+class TracingExactlyOnceAckHandler
+    : public pubsub::ExactlyOnceAckHandler::Impl {
+ public:
+  explicit TracingExactlyOnceAckHandler(
+      std::unique_ptr<pubsub::ExactlyOnceAckHandler::Impl> child,
+      opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>
+          subscribe_span)
+      : child_(std::move(child)), subscribe_span_(std::move(subscribe_span)) {}
+
+  ~TracingExactlyOnceAckHandler() override = default;
+  future<Status> ack() override {
+    subscribe_span_->AddEvent("gl-cpp.message_ack");
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    opentelemetry::trace::StartSpanOptions options = RootStartSpanOptions();
+
+    options.kind = opentelemetry::trace::SpanKind::kInternal;
+    auto span = internal::MakeSpan(
+        subscription().subscription_id() + " ack",
```

Comment:
nit: We should only call `subscription()` once. Each one makes a copy of the whole object, when at times we only need to copy the subscription id.

```
auto const subscription = subscription();
```

Then use that below.